core/src/retransmit_stage.rs

```diff
@@ -131,45 +132,37 @@ impl RetransmitStats {
     }
 }
 
-// Map of shred (slot, index, type) => list of hash values seen for that key.
-type ShredFilter = LruCache<ShredId, Vec<u64>>;
+struct ShredDeduper<const K: usize> {
+    deduper: Deduper<K, [u8]>,
+    filter: Deduper<K, [u64]>,
+}
 
-// Returns true if shred is already received and should skip retransmit.
-fn should_skip_retransmit(
-    key: ShredId,
-    shred: &[u8],
-    shreds_received: &mut ShredFilter,
-    packet_hasher: &PacketHasher,
-) -> bool {
-    match shreds_received.get_mut(&key) {
-        Some(sent) if sent.len() >= MAX_DUPLICATE_COUNT => true,
-        Some(sent) => {
-            let hash = packet_hasher.hash_shred(shred);
-            if sent.contains(&hash) {
-                true
-            } else {
-                sent.push(hash);
-                false
-            }
-        }
-        None => {
-            let hash = packet_hasher.hash_shred(shred);
-            shreds_received.put(key, vec![hash]);
-            false
+impl<const K: usize> ShredDeduper<K> {
+    fn new<R: Rng>(rng: &mut R, false_positive_rate: f64, num_bits: u64) -> Self {
+        Self {
+            deduper: Deduper::new(rng, false_positive_rate, num_bits),
+            filter: Deduper::new(rng, false_positive_rate, num_bits),
         }
     }
-}
 
-fn maybe_reset_shreds_received_cache(
-    shreds_received: &mut ShredFilter,
-    packet_hasher: &mut PacketHasher,
-    hasher_reset_ts: &mut Instant,
-) {
-    const UPDATE_INTERVAL: Duration = Duration::from_secs(1);
-    if hasher_reset_ts.elapsed() >= UPDATE_INTERVAL {
-        *hasher_reset_ts = Instant::now();
-        shreds_received.clear();
-        packet_hasher.reset();
+    fn maybe_reset<R: Rng>(&mut self, rng: &mut R, reset_cycle: &Duration) {
+        self.deduper.maybe_reset(rng, reset_cycle);
+        self.filter.maybe_reset(rng, reset_cycle);
+    }
+
+    fn check_duplicate(&self, key: ShredId, shred: &[u8], max_duplicate_count: usize) -> bool {
+        if self.deduper.check_duplicate(shred) {
+            return true;
+        }
+        let (slot, index, shred_type) = key.unpack();
+        let mut key = [slot, u64::from(index), u64::from(u8::from(shred_type)), 0];
+        for k in 0..max_duplicate_count {
+            key[3] = k as u64;
```

Comment:
this could use some documentation.  or better yet, a new type with self-descriptive field names

Reply:
Changed this to `(ShredId, i)` and added comment.